### ISSUE TITLE
[fcos] questions about the interface between host and container runtime

### ISSUE THREAD

**Issue by dustymabe** [boot2podman/boot2podman#17]
In https://github.com/coreos/fedora-coreos-tracker/issues/231 we had a discussion about the use of varlink on Fedora CoreOS as a remote interface for container management. We'd like to gain a better understanding of how the boot2podman project is planning to set up and use this so that we know about this in the future when we make policy decisions about the host. Some questions that come to mind:

- Will you be using ssh as an authentication and transport mechanism?
- If using SSH will you be providing an ssh key via ignition?
- Will you be using the `podman remote` command at all?

In general we're trying to discourage the use of SSH on FCOS, but understand that some tools have been built to assume it exists and are trying to keep track of places we know it is being used for bridging functionality. 

cc @praveenkumar @lucab

**Comment by bgilbert** [comment]
Also, if you'll be using SSH and specifying your own key:
- Will you use the `command` keyword in `authorized_keys` so the SSH key can only run varlink?
- Will you create a separate FCOS user that doesn't have sudo access?

**Comment by afbjorklund** [comment]
Hi, thanks for reaching out regarding podman and varlink. I think the current set up with the resolver and the podman.sock is OK, and there is some more documentation on how it is supposed to be used the sister project to `boot2podman`, `podman-machine`: https://github.com/boot2podman/machine

> We'd like to gain a better understanding of how the boot2podman project is planning to set up and use this so that we know about this in the future when we make policy decisions about the host. 

_Some answers to questions:_

>    Will you be using ssh as an authentication and transport mechanism?

Yes. I don't think varlink offers any secure alternatives to this ? (like HTTPS)

Not completely sold of the benefit of using varlink, compared to `ssh sudo podman`...

>    If using SSH will you be providing an ssh key via ignition?

Sortof. The machine allocation has a solution similar to ignition (I think?) where the ssh keys are shipped as a side container to the initial image. Currently it is using a tar-based approach to this:

https://github.com/boot2podman/machine/blob/master/libmachine/mcnutils/b2p.go#L473_L528

https://github.com/boot2podman/boot2podman/blob/master/files/init.d/autoformat#L114_L117

>    Will you be using the podman remote command at all?

Was using pypodman, but I suppose this can be replaced with podman-remote when ready.

>    Will you use the command keyword in authorized_keys so the SSH key can only run varlink?

The boot2podman distribution is similar to boot2docker in that the default user has both ssh access and sudo access. Currently using root user (`ssh root@`) for accessing the varlink socket.

>    Will you create a separate FCOS user that doesn't have sudo access?

Currently **not** using FCOS, but regular Fedora. And the root user is used, rather than sudo.
That is: varlink uses root@, but podman is also available with sudo or rootless using tc@

Note: "tc" is the default username for Tiny Core, it is used instead of the previous "docker"
The Fedora image uses the same username in order to work properly with podman-machine.

https://github.com/boot2podman/boot2podman-fedora-iso/blob/master/fedora.template#L94_L101

No particular plans for FCOS due to limited understanding, but can book a meeting perhaps ?

**Comment by lucab** [comment]
> No particular plans for FCOS due to limited understanding, but can book a meeting perhaps ?

For context, this ticket is coming from an explicit request at https://github.com/coreos/fedora-coreos-tracker/issues/231#issuecomment-520840937.

**Comment by dustymabe** [comment]
> Hi, thanks for reaching out regarding podman and varlink. I think the current set up with the resolver and the podman.sock is OK, and there is some more documentation on how it is supposed to be used the sister project to `boot2podman`, `podman-machine`: https://github.com/boot2podman/machine

No problem. Thanks for the answers!

> 
> > Will you be using ssh as an authentication and transport mechanism?
> 
> Yes. I don't think varlink offers any secure alternatives to this ? (like HTTPS)

I believe varlink can talk over any remote bridging mechanism. You just have to bridge the listening socket to a transport (ssh, plain tcp/ip, websocket, etc). It would require you to set this up when you bring up the virtual machine. SSH is there by default and doesn't require any special set up. 

> 
> Not completely sold of the benefit of using varlink, compared to `ssh sudo podman`...

I think the idea is that it offers access to an API and thus can be integrated into other applications (i.e. like a GUI, or a component of a larger application like an IDE). 

> 
> > If using SSH will you be providing an ssh key via ignition?
> 
> Sortof. The machine allocation has a solution similar to ignition (I think?) where the ssh keys are shipped as a side container to the initial image. Currently it is using a tar-based approach to this:
> 
> https://github.com/boot2podman/machine/blob/master/libmachine/mcnutils/b2p.go#L473_L528
> 
> https://github.com/boot2podman/boot2podman/blob/master/files/init.d/autoformat#L114_L117
> 

Thanks for the links.

> > Will you be using the podman remote command at all?
> 
> Was using pypodman, but I suppose this can be replaced with podman-remote when ready.

pypodman works. I wasn't trying to advocate for `podman-remote` just was interested in exactly how you were communicating

> 
> > Will you use the command keyword in authorized_keys so the SSH key can only run varlink?
> 
> The boot2podman distribution is similar to boot2docker in that the default user has both ssh access and sudo access. Currently using root user (`ssh root@`) for accessing the varlink socket.

We were interested to know if you'd be creating a user that could only speak to podman (via varlink) and do nothing else. It looks like you plan on using the generic user that is already created and used for general ssh access as well. 

> 
> > Will you create a separate FCOS user that doesn't have sudo access?
> 
> Currently **not** using FCOS, but regular Fedora. And the root user is used, rather than sudo.
> That is: varlink uses root@, but podman is also available with sudo or rootless using tc@
> 
> Note: "tc" is the default username for Tiny Core, it is used instead of the previous "docker"
> The Fedora image uses the same username in order to work properly with podman-machine.
> 
> https://github.com/boot2podman/boot2podman-fedora-iso/blob/master/fedora.template#L94_L101
> 
> No particular plans for FCOS due to limited understanding, but can book a meeting perhaps ?


Thanks for helping us gain some understanding.

**Comment by afbjorklund** [comment]
Here are the parts that I consider part of the interface, or the "contract", between host and machine:

1) Access to login using ssh (key-based, not password) and running the command `varlink bridge`

2) Access to the socket `/run/org.varlink.resolver` - the resolver service, open to anyone (`0666`)

3) Access to the socket `/run/podman/io.podman` - the podman service, open to `root` only (`0600`)

This is implemented in _different_ ways in the different implementations, but both provide these at boot...

The `boot2podman.iso` (tinycore) uses regular init.d shell scripts:

/usr/local/etc/init.d/varlink
/usr/local/etc/init.d/podman

The `boot2podman-fedora.iso` uses socket-activated systemd units:

/usr/lib/systemd/system/com.redhat.resolver.service
/usr/lib/systemd/system/org.varlink.resolver.socket
/usr/lib/systemd/system/io.podman.service
/usr/lib/systemd/system/io.podman.socket

----

Then it can be accessed in the same way, from any varlink client:

```console
$ varlink --bridge="$VARLINK_BRIDGE" info
Interfaces:
  io.podman
  org.varlink.resolver

$ varlink --bridge="$VARLINK_BRIDGE" call org.varlink.resolver.Resolve '{"interface": "io.podman"}'
{
  "address": "unix:/run/podman/io.podman;mode=0600"
}
$ varlink --bridge="$VARLINK_BRIDGE" call io.podman.GetVersion
{
  "built": "1969-12-31T19:00:00-05:00",
  "git_commit": "",
  "go_version": "go1.12.5",
  "os_arch": "linux/amd64",
  "remote_api_version": 1,
  "version": "1.4.2"
}
```
The resolver is just to avoid hardcoding `/run/podman/io.podman`.

For some reason upstream prefers hardcoding, but I guess if you only have one service anyway - you might as well hardcode /run/podman/io.podman rather than hardcoding /run/org.varlink.resolver ? Like a DNS server that only knows one host, might as well skip the lookup and use host address directly... I thought I was doing it "by the book", but it seems like nobody else is reading the same one anyway ?

* https://www.projectatomic.io/blog/2018/05/podman-varlink/
* https://podman.io/blogs/2019/01/16/podman-varlink.html

> We were interested to know if you'd be creating a user that could only speak to podman (via varlink) and do nothing else. It looks like you plan on using the generic user that is already created and used for general ssh access as well.

As far as I know, the main podman service **requires** root access. So what I did was that I opened up access for the default user with sudo access, to use the same credentials to login as root directly. So I am reusing the key, not the user.

```sh
cp /home/tc/.ssh/authorized_keys /root/.ssh/
chmod 600 /root/.ssh/authorized_keys
```

Not using varlink for rootless podman.
